crates/holochain/src/core/state/cascade.rs

```diff
@@ -58,20 +55,106 @@ mod test;
 
 pub mod error;
 
-pub struct Cascade<'a, Network = HolochainP2pCell, MetaVault = MetadataBuf, MetaCache = MetadataBuf>
+/////////////////
+// Helper macros
+/////////////////
+
+/// Get an item from an option
+/// or return early from the function
+macro_rules! get_or_return {
+    ($n:expr) => {
+        match $n {
+            Some(n) => n,
+            None => return Ok(()),
+        }
+    };
+    ($n:expr, $ret:expr) => {
+        match $n {
+            Some(n) => n,
+            None => return Ok($ret),
+        }
+    };
+}
+
+/// Return from a function if
+/// an item is found otherwise continue
+macro_rules! return_if_found {
+    ($i:expr) => {
+        if let Some(i) = $i {
+            return Ok(Some(i));
+        }
+    };
+}
+
+/// Search every level that the cascade has been constructed with
+macro_rules! search_all {
+    ($cascade:expr, $fn:ident, $hash:expr) => {{
```

Comment:
Pretty sure if you change the fn to be an `expr`, you'll be able to use a closure in its place

```suggestion
    ($cascade:expr, $fn:expr, $hash:expr) => {{
```

Reply:
Nah can't use closures with generics